### DISCUSSION TITLE
Does packer support nested lists of plugins?

### DISCUSSION THREAD

**Discussion by danielo515** [wbthomason/packer.nvim#1026] [General]
Hello, and as always, thanks again for packer.
Today I realized that, under my main list of plugins, I have some nested lists of plugins, and I am not sure if packer is able to pick them? Maybe it is some side effect from a flatten call, or maybe it is an intended design decission? The thing is that I don't get any error, and everything seems to work, but I didn't saw any reference to poviding nested list of plugins.
To put it in concrete terms, this is what I am talking about

```lua
use {
  "plugin/x", -- normal plain usage
  { "plugin/y" }, -- plugin defined as a table
  {
    { "plugin/a"}, -- nested  table of plugins defined as tables
    { "plugin/b"},
  }

}

**Comment by EdenEast** [comment]
Yes packer does support nested lists of plugins. A plugin `spec` can either be a `string` or a `table`. A string is just a simple plugin with no extra configuration. A table is a `spec` that has some sort of additional keys assigned (`config`, `opt`, `ft`...). A table spec MUST have the first value as the name as a string. For example this is invalid:

```lua
use {
  { config = function() end, "user/repo"  }, -- invalid
}

-- All valid
use {
  "user/repo", -- single
  {
    "user/repo", -- list of singles
    "user/repo",
  },
  { "user/repo" }, -- complex with additional keys
  {
    { "user/repo" }, -- list of complex with additional keys
    { "user/repo" },
    { "user/repo" },
  },
}

```

This is the same for the `requires` key's value. Make sure you have the [correct number of tables](https://github.com/wbthomason/packer.nvim/issues/1000#issuecomment-1214617368).

**Reply by danielo515** [reply]
That's good to know, thanks. My last question, specifically about the list of complex with additional keys... is the allowed nesting arbitrary? Or just one level of nesting is permitted? In case of more than the allowed nesting is provided, will it error out?

Thank you very much for your quick answer by the way.

**Reply by EdenEast** [reply]
There is no restriction on the level of nesting. When nesting you have to make sure that the each nesting level is a list.

**Comment by danielo515** [comment]
Awesome! Thank you so much for the answers.

On Tue, Aug 23, 2022 at 3:47 PM EdenEast ***@***.***> wrote:

> There is no restriction on the level of nesting. When nesting you have to
> make sure that the each nesting level is a list.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wbthomason/packer.nvim/discussions/1026#discussioncomment-3456922>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARKJWJ2IHG2KJVAQ4GZ7XDV2TI7PANCNFSM57G3BZGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
---
https://danielorodriguez.com
